torch_xla/csrc/runtime/pjrt_compilation_client.cc

```diff
@@ -0,0 +1,707 @@
+#include "torch_xla/csrc/runtime/operation_manager.h"
+#include "torch_xla/csrc/runtime/pjrt_registry.h"
+#include "torch_xla/csrc/runtime/profiler.h"
+#include "torch_xla/csrc/runtime/stablehlo_helper.h"
+#include "torch_xla/csrc/runtime/tensor_source.h"
+#include "torch_xla/csrc/runtime/tf_logging.h"
+#include "torch_xla/csrc/runtime/xla_coordinator.h"
+#include "torch_xla/csrc/thread_pool.h"
+#include "tsl/profiler/lib/traceme.h"
+#include "xla/client/xla_builder.h"
+#include "xla/client/xla_computation.h"
+#include "xla/layout_util.h"
+#include "xla/literal.h"
+#include "xla/pjrt/pjrt_c_api_client.h"
+#include "xla/pjrt/pjrt_client.h"
+#include "xla/pjrt/pjrt_executable.h"
+#include "xla/protobuf_util.h"
+#include "xla/shape.h"
+
+using xla::internal::XlaBuilderFriend;
+
+namespace torch_xla {
+namespace runtime {
+
+namespace {
+
+// Builds a map from the device's global ordinal to its index in the `devices`
+// array.
+std::unordered_map<int, int> build_index_map(
+    const std::vector<std::string>& devices) {
+  std::unordered_map<int, int> device_index;
+  for (int i = 0; i < devices.size(); ++i) {
+    std::vector<std::string> device_spec = absl::StrSplit(devices[i], ':');
+    XLA_CHECK_EQ(device_spec.size(), 2)
+        << "Invalid device specification: " << devices[i];
+    int global_ordinal = std::stoi(device_spec[1]);
+    device_index[global_ordinal] = i;
+  }
+  return device_index;
+}
+
+// Builds the xla::Shape of the output xla::Literal on the host.
+xla::Shape host_output_shape(xla::PjRtBuffer* buffer) {
+  xla::Shape shape = xla::ShapeUtil::MakeShape(
+      buffer->element_type(), buffer->logical_dimensions().value());
+  *shape.mutable_layout() = xla::GetXlaLayoutUnsafe(buffer->layout());
+
+  return xla::ShapeUtil::DeviceShapeToHostShape(shape);
+}
+
+torch::lazy::hash_t hash_comp_env() {
+  // TODO(piz): since the client is nullptr, we can't retrive all information
+  // like PjRtComputationClient. Think about a way to construct the hashing.
+  torch::lazy::hash_t hash = hash::HashXlaEnvVars();
+  return hash;
+}
+
+}  // namespace
+
+std::string PjRtCompilationClient::PjRtDeviceToString(
+    xla::PjRtDevice* const device) const {
+  std::string platform =
+      absl::AsciiStrToUpper(device->client()->platform_name());
+  int ordinal = global_ordinals_.at(device->id());
+  std::string str = absl::StrFormat("%s:%d", platform, ordinal);
+  return str;
+}
+
+std::vector<std::string> PjRtCompilationClient::PjRtDevicesToString(
+    absl::Span<xla::PjRtDevice* const> devices) const {
+  std::vector<std::string> strs;
+  strs.reserve(devices.size());
+
+  for (auto* device : devices) {
+    strs.push_back(PjRtDeviceToString(device));
+  }
+
+  return strs;
+}
+
+PjRtCompilationClient::PjRtCompilationClient(
+    std::string& virtual_topology_str) {
+  std::string device_type = sys_util::GetEnvString(env::kEnvPjRtDevice, "");
+
+  // With compilation only, client_ and coordinator_ are nullptr. The point we
```

Comment:
It doesn't seem right to call InitializePjrt and throw away the results. If you're planning to hardcode this for only TPU support, just put `pjrt::LoadPjrtPlugin("tpu", etc)` here and skip everything else. (Also, the branch you added over in `pjrt_registry` won't fire by default because dynamic plugins are enabled by default now)

If you want to support general devices, then you would need to add a new function in `pjrt_registry` like `InitializeTopology` that loads the plugin and returns the topology.

Reply:
That's why I mentioned ` XLA_REGISTER_INSTALLED_PLUGINS=0` in the PR description to skip the dynamic plug in.

Simply to use `pjrt::LoadPjrtPlugin("tpu", etc)` seems to be a good solution. Let me address that.